### DISCUSSION TITLE
Can I allow my users to point their custom domains to subdomains of my app?

### DISCUSSION THREAD

**Discussion by mikealche** [vercel/next.js#12751] [Help] [answered]
Hello there! 

First of all thanks for the amazing project! 
I'm currently building a site that allows people to create their own landing pages. 

I'd like to be able to allow my users to point their custom domains like 
www.user-1.com to my app.

Right now my users find their landing pages at 
www.myapp.com/user-page-slug

and I'd like to enable them to access them at:

www.user-1.com/user-page-slug

Can you recommend me how to do that? 

Thanks a lot!

**Comment by juhaelee** [comment]
@miguelalche were you able to come up with a solution for this?

**Reply by steven-tey** [reply]
@juhaelee I just posted an example for this in my comment below, hope it helps!

**Comment by wyeo** [comment]
@miguelalche i think this link can help you : [https://vercel.com/blog/wildcard-domains](https://vercel.com/blog/wildcard-domains)

**Comment by mikealche** [comment]
@wyeo as I see, right now the example uses the window.location host in order to identify the client. 

So If I want to give my clients the ability to point their custom domains (like www.enterpriseclient.com) to (client1.mynextapp.com) everything would break, right?

**Reply by chesampat** [reply]
@mikealche Did you find your answer? Currently working this out myself now!

**Reply by mikealche** [reply]
Hey! I think you would need to ask the clients for the custom domain that they will be using and store it in a db. Then when a request comes in, you can check if the domain is in the list of domains you have and get back the correct info :)

**Reply by steven-tey** [reply]
@kareemcoding I just posted an example for this in my comment below, hope it helps!

**Comment by jaxomlotus** [comment]
@mikealche  Also looking for a solution to this... I want to enable to my customers to just redirect their own domains to a specific url in my app, associated with their account. I'm not seeing a way to do this inside Vercel without running into SSL issues. And because of the nature of my app there would be too many to store those domains inside Vercel via API or manually. It would have to be dynamic. Any solutions are welcome!

**Reply by steven-tey** [reply]
@jaxomlotus I just posted an example for this in my comment below, hope it helps!

**Comment by eunjae-lee** [comment]
Also looking for an answer here.

Let's say I can configure sub domains using wildcard like mentioned above:

user-1.mysaas.com

but when this user wants to set up their own domain, what needs to happen? On this domain registrar, should they need to put vercel's dns? How does vercel know that domain needs to point my application?
I haven't done such thing ever, so I just lack the general knowledge of how it's supposed to work. I appreciate any advice.

**Reply by jaxomlotus** [reply]
Have you looked at Vercel's domain api documentation? This is covered.

You will need to configure your users domain inside vercel, either manually or via the API.

**Reply by SleepWalker** [reply]
Here is one more discussion with updates regarding next@10.2 and new redirect features:

https://github.com/vercel/next.js/discussions/12848#discussioncomment-680225

**Reply by ToledoNicola** [reply]
I want to do the same using wildcard , So from what I understand i have to:
1 get The user customizzazione domain es. pippo.com
2 add pippo.com to my vercel project domains using this api (https://vercel.com/docs/rest-api#endpoints/domains/add-a-new-domain)
3 The user have to set in his domain register The cname to point my server
4 in my next.js app I have to create a dynamic indexes. pages/[host].tsx and redirect any wild art to that page  using this approach (https://github.com/vercel/next.js/discussions/12848#discussioncomment-680225)
it's right? Thanks 🙂

**Reply by eunjae-lee** [reply]
For anyone interested, I wrote two articles afterwards:

https://buttondown.email/quill/archive/subdomain-and-custom-domain/

https://buttondown.email/quill/archive/subdomain-and-custom-domain-part-2/

**Reply by ToledoNicola** [reply]
interesting, but have you tried the `rewrites()` solution? 
because with that you can use the `getStaticProps()` and ISR

also I'd like to see the workflow in CircleCI  if possible 🙂

**Reply by eunjae-lee** [reply]
I guess you're referring to [this](https://vercel.com/support/articles/can-i-redirect-from-a-subdomain-to-a-subpath?query=rewrite#configuring-a-rewrite-with-'has'-field)? Wow, I didn't know about it so far. I think that's the correct solution. My CircleCI workflow isn't pretty.

1. I call CircleCI api with some parameters (like subdomain)
2. CircleCI runs a script
3. That script clones my repo, and with parameters as environment variables, it deploys a new app on Vercel (using Vercel CLI.) and configures the subdomain.
4. My app monitors the CircleCI task by periodically pinging the task status. When it's done, confetti.

The biggest downside of this approach is, whenever I need an update, I need to re-deploy all the apps subdomain by subdomain. If it's just a single app with rewrites() solution, that would've been much better. I should give it a try.

**Reply by steven-tey** [reply]
@eunjae-lee I just posted an example for this in my comment below, hope it helps!

As for the API to add custom domains, we will be adding that example very soon in the early access program, feel free to join that as well!

**Reply by eunjae-lee** [reply]
@steven-tey yeah, since the last comment by @ToledoNicola , I tried it out and

```js
rewrites() {
    return [
      {
        source: "/:path*",
        has: [
          {
            type: "host",
            value: "(?<host>.*)",
          },
        ],
        destination: "/:host/:path*",
      },
    ]
  },
```

this enabled it. Now I have `pages/[host]/[...slug].tsx` file that covers it all :)

**Reply by steven-tey** [reply]
@eunjae-lee! Awesome! This works as well but we're encouraging folks to do rewrites with Edge Functions ([example](https://github.com/vercel/examples/tree/main/edge-functions/hostname-rewrites)) instead of doing it inside `next.config.js` for scalability purposes!

**Reply by eunjae-lee** [reply]
@steven-tey Thanks for the reply. I'm reading the repo right now but could  you help me understand on how it could be better in terms of scalability?

does that approach also enable Incremental Static Generation?

**Reply by steven-tey** [reply]
@eunjae-lee Sure! With the `next.config.js` method, there is a limit as to how many rewrites you can perform, but with Edge Functions, there is no limit at all. And yes, you can do ISR with the Edge Functions example that I shared above! :)

**Reply by eunjae-lee** [reply]
@steven-tey ah alright, but as you see there, I have a wildcard, which means I only need one rewrite rule :) 
But I will keep an eye on the edge functions which is definitely useful in many cases!

**Comment by steven-tey** [comment]
Hey @mikealche! Great question – we recently rolled out a [simple example](https://github.com/vercel/examples/tree/main/edge-functions/hostname-rewrites) that allows you to create unique content pages with a multi-tenant infrastructure using [edge function](https://vercel.com/features/edge-functions) rewrites and [ISR](https://vercel.com/docs/concepts/next.js/incremental-static-regeneration). Each of your users gets assigned a unique subdomain when they create their account, with the (usually paid) option to add a custom domain. 

We're also rolling out more complex boilerplates (e-commerce storefront builders, website builders) + supplementary code snippets (static tweets, domains API, OG image generator) in an [early access program](https://eap.vercel.sh/) – feel free to join if you're interested (click "sign up" on the Platforms on Vercel program).

**Comment by steven-tey** [comment, marked as answer]
Quick update on this – we recently launched the Platforms Starter Kit that's a comprehensive template for building multi-tenant applications with built-in custom domains support.: https://vercel.com/blog/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Reply by steven-tey** [reply]
CC: @mikealche @juhaelee @eunjae-lee @jaxomlotus

**Reply by wyeo** [reply]
Amazing ! Great job @steven-tey 👌🏿

**Reply by hectorruizlaph** [reply]
Hi there, is there a way to add the functionality that users can link their own domain. For example in my app a user has the content in hector.mysite.com and I want the same content to be routed to their external domain, example hectorsite.com Is there a way to make this happen? can someone point a blog or something useful I can read to make this happen? stack: vercel, nextjs, supabase. @steven-tey

**Reply by johannbuscail** [reply]
Hey @hectorruizlaph ! Have you found a solution ?

**Reply by iamkoby** [reply]
platforms-starter-kit still doesn't let you change css per domain

**Comment by jaxomlotus** [comment]
My team built this functionality before, but need to dig through old code
to remember how we did it. It's definitely possible.
